diff --git a/openmax/vdec/src/omx_vdec_aw_decoder_linux.c b/openmax/vdec/src/omx_vdec_aw_decoder_linux.c
index 8a619ca..a01f3b4 100644
--- a/openmax/vdec/src/omx_vdec_aw_decoder_linux.c
+++ b/openmax/vdec/src/omx_vdec_aw_decoder_linux.c
@@ -72,6 +72,11 @@ typedef struct OmxAwDecoderContext
 
     OMX_BOOL               bUseZeroCopyBuffer;
 
+    char*                  pYV12;
+    VideoPicture           CvtPicture;
+
+    int                    tWidth;
+    int                    tHeight;
 
 }OmxAwDecoderContext;
 
@@ -403,6 +408,52 @@ static void liReopenVideoEngine(OmxAwDecoderContext *pCtx)
     return ;
 }
 
+static int yv12_align_fill(char* p_dst, int w, int h, int aw, int ah)
+{
+	int i, j;
+	int map = aw * ah;
+	int v_size = (map) >> 2;
+	char *pv = p_dst + map;
+	char *pu = pv + v_size;//  / 4
+
+	int dw = aw - w;
+	int dh = ah -h;
+	int dw_h = dw / 2;
+	int dh_h = dh / 2;
+	int w_h = w / 2;
+	int h_h = h / 2;
+	int aw_h = aw / 2;
+	int ah_h = ah / 2;
+	int map_oft = aw_h * h_h;
+	
+	if((w == aw) && (h == ah))
+		return 0;
+
+
+	for(i=0;i<h_h;i++)
+	{
+		for(j=0;j<dw_h;j++)
+		{
+			pv[w_h + j + aw_h * i] = 128;
+			pu[w_h + j + aw_h * i] = 128;
+		}
+
+	}
+	
+
+	for(i=0;i<dh_h;i++)
+	{
+		for(j=0;j<aw_h;j++)
+		{
+			pv[map_oft + j + aw_h * i] = 128;
+			pu[map_oft + j + aw_h * i] = 128;
+		}
+		
+	}
+
+	return 0;
+}
+
 static int __liCallback(OmxDecoder* pDec, DecoderCallback callback,
                                     void* pUserData)
 {
@@ -471,7 +522,8 @@ static int __liPrepare(OmxDecoder* pDec)
 
     pCtx->mVideoConfig.nAlignStride       = pCtx->mGpuAlignStride;
 
-    pCtx->mVideoConfig.eOutputPixelFormat = PIXEL_FORMAT_YV12;//* Used to be YV12.
+    //pCtx->mVideoConfig.eOutputPixelFormat = PIXEL_FORMAT_YV12;//* Used to be YV12.
+    pCtx->mVideoConfig.eOutputPixelFormat = PIXEL_FORMAT_YUV_MB32_420;
 
 #if (ENABLE_SCALEDOWN_WHEN_RESOLUTION_MOER_THAN_1080P)
     if (pCtx->mStreamInfo.nWidth > 1920
@@ -483,6 +535,11 @@ static int __liPrepare(OmxDecoder* pDec)
     }
 #endif
 
+        pCtx->tWidth = pCtx->mStreamInfo.nWidth;
+        pCtx->tHeight = pCtx->mStreamInfo.nHeight;
+        pCtx->mStreamInfo.nWidth = (pCtx->mStreamInfo.nWidth + 31) & ~31;
+        pCtx->mStreamInfo.nHeight = (pCtx->mStreamInfo.nHeight + 31) & ~31;
+
     if(pCtx->mStreamInfo.eCodecFormat == VIDEO_CODEC_FORMAT_WMV3)
     {
         OMX_LOGD("*** pCtx->mStreamInfo.bIsFramePackage to 1 when it is vc1");
@@ -503,6 +560,14 @@ static int __liPrepare(OmxDecoder* pDec)
     pCtx->decMemOps  = pCtx->mVideoConfig.memops;
     CdcMemOpen(pCtx->decMemOps);
     pCtx->mStreamInfo.bIsFramePackage = 1;
+    pCtx->pYV12 = (char*)malloc(sizeof(char) * pCtx->mStreamInfo.nWidth * pCtx->mStreamInfo.nHeight * 3 / 2);
+if(pCtx->pYV12 == NULL)
+{
+    loge("pCtx->pYV12 malloc fail\n");
+    return -1;
+}
+pCtx->CvtPicture.pData0 = pCtx->pYV12;
+pCtx->CvtPicture.ePixelFormat = PIXEL_FORMAT_YV12;
     ret = InitializeVideoDecoder(pCtx->m_decoder,
                                  &(pCtx->mStreamInfo),
                                  &pCtx->mVideoConfig);
@@ -702,6 +767,8 @@ static OMX_BUFFERHEADERTYPE* __liDrain(OmxDecoder* pDec)
         logw("the pPicture is null when request displayer picture!");
         return NULL;
     }
+    RotatePicture(pCtx->decMemOps, pCtx->pPicture, &pCtx->CvtPicture, 0, 32, 32);
+    yv12_align_fill(pCtx->CvtPicture.pData0, pCtx-> tWidth, pCtx-> tWidth, pCtx->mStreamInfo.nWidth, pCtx->mStreamInfo.nHeight);
     logv("*** picture info: w(%d),h(%d),offset,t(%d),b(%d),l(%d),r(%d)",
          pCtx->pPicture->nWidth,    pCtx->pPicture->nHeight,
          pCtx->pPicture->nTopOffset,pCtx->pPicture->nBottomOffset,
@@ -711,7 +778,7 @@ static OMX_BUFFERHEADERTYPE* __liDrain(OmxDecoder* pDec)
     OMX_PARAM_PORTDEFINITIONTYPE* outDef = getPortDef(pCtx->pOutPort);
 #if USE_ALIGN_SIZE
     AlignCopyYV12((unsigned char*)pOutBufHdr->pBuffer,
-                                (unsigned char*)pCtx->pPicture->pData0,
+                                (unsigned char*)pCtx->CvtPicture.pData0,
                                 outDef->format.video.nFrameWidth,
                                 outDef->format.video.nFrameHeight);
 
@@ -780,6 +847,13 @@ static  void __liClose(OmxDecoder* pDec)
     OmxTryPostSem(pCtx->mSemInData);
     OmxTryPostSem(pCtx->mSemOutBuffer);
     OmxReleaseMutex(pCtx->awMutexHandler);
+    
+    if(pCtx->pYV12 != NULL)
+    {
+        free(pCtx->pYV12);
+        pCtx->pYV12 = NULL;
+    }
+    
     if(pCtx->m_decoder != NULL)
     {
         DestroyVideoDecoder(pCtx->m_decoder);
@@ -835,7 +909,7 @@ static int __liSetOutputEos(OmxDecoder* pDec)
 static void __liGetFormat(OmxDecoder* pDec)
 {
     OmxAwDecoderContext *pCtx = (OmxAwDecoderContext*)pDec;
-    pCtx->mVideoConfig.eOutputPixelFormat = PIXEL_FORMAT_YV12;
+    pCtx->mVideoConfig.eOutputPixelFormat = PIXEL_FORMAT_YUV_MB32_420;
 }
 
 static inline void __liSetExtBufNum(OmxDecoder* pDec, OMX_S32 num)
